Detect only changed files in HEAD commit

handk85 opened the pull request:
get_differences API without beforeCommitSpecifier parameter returns the entire files in a repository. It introduces a lot of unnecessary updates in an S3 bucket. I added several lines of code to extract the files that are changed in the HEAD commit and update them only. My change requires extended permission than AWSLambdaExecute since the function need to update the environment variable, beforeCommitSpecifier, at the end of execution.

handk85 committed:
Detect only changed files in HEAD commit

michaelniedermayr commented:
Hi handk85, thx for your commit and sorry for my late response. You address a very good point. 
I just reviewed your changes and I have some questions. I understand that the lambda function based on your code just looks for changes and only updates that part. Afterwards the beforeCommitSpecifier is updated and stored as an enviroment varible. This however turns a stateless lambda in a stateful function which cotradicts the idea of lambda function per definition. Of course this is not a problem per se but I am worried that this could confuse people. 
I was also wondering how do you run the function for the first time when no beforeCommitSpecifier is defined.

handk85 committed:
Use SSM parameter to keep beforeCommitSpecifier instead of Lambda environment variable

handk85 commented:
Thanks for the comment. I agree with your point. It looks like cheating to use an environment variable to keep beforeCommitSpecifier. When I wrote the code, however, I thought using either S3 or DynamoDB to keep tracking a simple variable is too heavy. Fortunately, I found the [SSM parameter](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html) which seems like feasible solution to this scenario. Please check the updated change. You have to add AmazonSSMFullAccess permission to get/update the parameter.

handk85 committed:
remove unnecessary comment

handk85 committed:
correct wrong comment

michaelniedermayr commented:
I finally found some time to play with your changes. I haven't known about SSM parameters but they seem to be very handy in situations like this. Good to know.
I noticed one issue in your chances. When you run the lambda function for the first time, the beforeCommitSpecifier SSM parameter is set but the files in the branch are not deployed to the S3 bucket. When the next commit is pushed and the lambda function is triggered again, only the changes will be deployed. This means, those files that where no modified will never show up in the S3 bucket.
I think a modified get_blob_list function that either fetches the files from the last commit or the entire branch (if beforeCommitSpecifier is not defined yet) could solve that issue. What do you think?

handk85 committed:
if beforeCommitSpecifier is not set, the entire files will be saved in S3 bucket

handk85 commented:
Thanks for reviewing. I added get_entire_blob_list function to retrieve the entire file list. Now, the lambda function stores the entire files in the repo to S3 bucket. In addition, I added repository prefix to SSM Parameter since the lambda function can be used for multiple repos.

michaelniedermayr commented:
Thanks for adjusting the code and your effort. I reviewed it again and it looks good. I made a few more changes to make it backward compatible and already pushed the new code.
